### DISCUSSION TITLE
There is no Suspense boundary in `next/dynamic`

### DISCUSSION THREAD

**Discussion by duannx** [vercel/next.js#74132] [Help]
### Summary

Hi, I tried to understand the behavior of `next/dynamic`. According to the [documentation](https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading), it uses `Suspense` and `React.lazy` under the hood. However, when I inspected the components using the React DevTools, I couldn't find any `Suspense`. I tested this with both `next@13.4.2` and `next@15.0.3` (page router). All I could see was a `LoadableComponent`.

```ts
const Test = dynamic(import('@/components/Test'))
```

![image](https://github.com/user-attachments/assets/7aa51d81-de1d-47a6-a34b-73b501a13e30)

The reason I asked this is that I am facing a hydration error when using `next/dynamic`. However, the error disappears when I use `React.lazy` with `Suspense`.

Thanks in advance!

**Comment by icyJoseph** [comment]
uhh the docs have not been properly updated!

- https://github.com/vercel/next.js/pull/67014

It was listed here, https://nextjs.org/blog/next-15-rc2#other-changes:

- [Breaking] next/dynamic: the deprecated suspense prop has been removed and when the component is used in the App Router, it won't insert an empty Suspense boundary anymore ([PR](https://github.com/vercel/next.js/pull/67014))
- [Breaking] Disallow using ssr: false option with next/dynamic in Server Components ([PR](https://github.com/vercel/next.js/pull/70378))

I'll try to make a PR to update the docs for App Router.

**Reply by duannx** [reply]
Hi @icyJoseph, thank you for the quick and informative reply. 
However, I encountered the problem with the **page router** and `next@13.4.2` as well. My project is quite old and still uses that version of Next.js

**Reply by icyJoseph** [reply]
Did you use the `suspense` prop? Since this is a bit unexpected, I won't make any assumptions

**Reply by duannx** [reply]
I didn’t. I tested it on a fresh project created with `npx create-next-app@13` (`next@13.5.7`) and got the same behavior.
```tsx
const ComponentA = dynamic(() => import("@/components/ComponentA"));
const ComponentB = dynamic(() => import("@/components/ComponentB"), {
  suspense: true,
});
const ComponentC = dynamic(() => import("@/components/ComponentC"), {
  suspense: false,
});
```
![image](https://github.com/user-attachments/assets/4675a57f-3499-4c9a-b61a-ca1e71c524da)

**Reply by icyJoseph** [reply]
Ok, could you try also adding a `loading` option? 

```js
const ComponentB = dynamic(() => import("@/components/ComponentB"), {
  suspense: true,
  loading: () => <p>loading...</p>
});
```

**Reply by duannx** [reply]
I added the loading option, but it still shows the same behavior.
```tsx
const ComponentA = dynamic(() => import("@/components/ComponentA"), {
  loading: () => <p>loading...</p>
});
const ComponentB = dynamic(() => import("@/components/ComponentB"), {
  suspense: true,
  loading: () => <p>loading...</p>
});
const ComponentC = dynamic(() => import("@/components/ComponentC"), {
  suspense: false,
  loading: () => <p>loading...</p>
});
```
![image](https://github.com/user-attachments/assets/ba6924f3-caef-4d81-ba5b-a32a1be445c1)

---
I even added a delay function to force the loading component to show, but there’s still no Suspense.
```tsx
function delay(ms: number) {
  if (typeof window === "undefined") return;
  return new Promise((res) => {
    setTimeout(res, ms);
  });
}

const ComponentA = dynamic(
  async () => {
    await delay(100000);
    return await import("@/components/ComponentA");
  },
  {
    loading: () => <p>loading...</p>,
  }
);
const ComponentB = dynamic(
  async () => {
    await delay(100000);
    return await import("@/components/ComponentB");
  },
  {
    suspense: true,
    loading: () => <p>loading...</p>,
  }
);
const ComponentC = dynamic(
  async () => {
    await delay(100000);
    return await import("@/components/ComponentC");
  },
  {
    suspense: false,
    loading: () => <p>loading...</p>,
  }
);
```

![image](https://github.com/user-attachments/assets/85429334-67b1-41a4-8912-6320afb54a38)

**Reply by duannx** [reply]
@icyJoseph Do you have any idea about this?
